tapir/shifts/models.py

```diff
@@ -1277,3 +1303,109 @@ class Meta:
                 name="shift_watch_constraint",
             )
         ]
+
+
+class ShiftRecurringWatchTemplate(models.Model):
+    """
+    class to generate recurring ShiftWatches from.
+    """
+
+    user = models.ForeignKey(
+        TapirUser, on_delete=models.CASCADE, related_name="shift_recurring_watch"
+    )
+    shift_templates = models.ManyToManyField(ShiftTemplate, blank=True)
+
+    shift_template_group = ArrayField(
+        models.CharField(
+            max_length=255,
+        ),
+        blank=True,
+        default=list,
+    )
+
+    weekdays = ArrayField(
+        models.IntegerField(blank=True, null=True, choices=WEEKDAY_CHOICES),
+        blank=True,
+        default=list,
+    )
+    staffing_status = ArrayField(
+        models.CharField(max_length=30, choices=get_staffingstatus_choices),
+        blank=False,
+        default=get_staffingstatus_defaults,
+    )
+
+    def create_shift_watches(self):
+        shifts_to_watch = set()
+
+        # scenario A: only weekdays given
+        if (
+            self.weekdays
+            and not self.shift_template_group
+            and not self.shift_templates.exists()
+        ):
+            for weekday in self.weekdays:
+                shifts = Shift.objects.filter(start_time__iso_week_day=(weekday + 1))
+                shifts_to_watch.update(shifts)
+
+        # scenario B: only shift_template_group given
+        elif (
+            not self.weekdays
+            and self.shift_template_group
+            and not self.shift_templates.exists()
+        ):
+            for category in self.shift_template_group:
+                shifts = Shift.objects.filter(
+                    shift_template__group__name__in=self.shift_template_group
+                )
+                shifts_to_watch.update(shifts)
+
+        # scenario C: both weekdays and shift_template_group given
+        elif self.weekdays and self.shift_template_group:
+            for weekday in self.weekdays:
+                for category in self.shift_template_group:
+                    shifts = Shift.objects.filter(
+                        start_time__iso_week_day=(weekday + 1),
+                        shift_template__group__name=category,
+                    )
+                    shifts_to_watch.update(shifts)
+
+        # scenario D: only shift_templates given
+        elif self.shift_templates.exists():
+            for template in self.shift_templates.all():
+                shifts = Shift.objects.filter(
+                    shift_template=template,
+                )
+                shifts_to_watch.update(shifts)
+
+        for shift in shifts_to_watch:
+            if not ShiftWatch.objects.filter(user=self.user, shift=shift).exists():
+                ShiftWatch.objects.create(
```

Comment:
We can optimize a lot here: we could get all the already watched shifts in a set outside of the `for`, then create watches only for `shifts_to_watch.difference(already_watched_shifts)`.

Also, the ShiftWatch could be done all at one in a `bulk_create`